packages/patternfly-4/react-core/src/components/Select/SelectToggle.js

```diff
@@ -0,0 +1,147 @@
+  /** The icon to display for the toggle. Defaults to CaretDownIcon. Set to null to not show an icon. */
+  iconComponent: PropTypes.func,
+  /** Additional props are spread to the container <button> */
+  '': PropTypes.any
+};
+
+const defaultProps = {
+  children: null,
+  className: '',
+  isExpanded: false,
+  parentRef: null,
+  isFocused: false,
+  isHovered: false,
+  isActive: false,
+  isPlain: false,
+  iconComponent: CaretDownIcon,
+  onToggle: Function.prototype
+};
+
+class SelectToggle extends Component {
+  componentDidMount() {
+    document.addEventListener('mousedown', this.onDocClick);
+    document.addEventListener('touchstart', this.onDocClick);
+    document.addEventListener('keydown', this.onEscPress);
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('mousedown', this.onDocClick);
+    document.removeEventListener('touchstart', this.onDocClick);
+    document.removeEventListener('keydown', this.onEscPress);
+  }
+
+  onDocClick = event => {
+    const { parentRef, isExpanded, onToggle, onClose } = this.props;
+    if (isExpanded && parentRef && !parentRef.contains(event.target)) {
+      onToggle && onToggle(false);
+      onClose && onClose();
+      this.toggle.focus();
+    }
+  };
+
+  onEscPress = event => {
+    const { parentRef, isExpanded, onToggle, onClose } = this.props;
+    if (
+      isExpanded &&
+      (event.key === 'Escape' || event.key === 'Tab') &&
+      parentRef &&
+      parentRef.contains(event.target)
+    ) {
+      onToggle && onToggle(false);
+      onClose && onClose();
+      this.toggle.focus();
+    }
+  };
+
+  onKeyDown = event => {
+    const { isExpanded, onToggle, onClose, onEnter } = this.props;
+    if (event.key === 'Tab' && !isExpanded) return;
+    event.preventDefault();
+    if ((event.key === 'Tab' || event.key === 'Enter' || event.key === ' ') && isExpanded) {
+      onToggle(!isExpanded);
+      onClose && onClose();
+    } else if ((event.key === 'Enter' || event.key === ' ') && !isExpanded) {
+      onToggle(!isExpanded);
+      onEnter();
+    }
+  };
+
+  render() {
+    const {
+      className,
+      children,
+      isExpanded,
+      isFocused,
+      isActive,
+      isHovered,
+      isPlain,
+      onToggle,
+      onEnter,
+      parentRef,
+      id,
+      type,
+      iconComponent: IconComponent,
+      ...props
+    } = this.props;
+    return (
+      <div
+        {...props}
+        id={id}
+        ref={toggle => {
+          this.toggle = toggle;
+        }}
+        className={css(
+          styles.selectToggle,
+          isFocused && styles.modifiers.focus,
+          isHovered && styles.modifiers.hover,
+          isActive && styles.modifiers.active,
+          isPlain && styles.modifiers.plain,
+          className
+        )}
+        tabIndex={0}
+        type={type || 'button'}
+        onClick={_event => onToggle && onToggle(!isExpanded)}
+        aria-expanded={isExpanded}
+        aria-haspopup="listbox"
```

Comment:
For this case, `aria-haspopup="listbox"` is correct, since the popup has `role="listbox"`. When the popup has `role="menu"` then this attribute can be either `aria-haspopup="true"` or `aria-haspopup="menu"` (i.e. the default popup role is a menu).

Reply:
sorry for the late reply, I had made this comment last week, but didn't hit the Comment button 😞